### PR TITLE
ocp13 replacement

### LINKED ISSUE #494: Upgrade ocp13 (registry) to Ubuntu 24.04 LTS
https://ubuntu.com/blog/canonical-releases-ubuntu-24-04-noble-numbat

These are running Ubuntu 20.04, which is EOL Apr 2025, with extended security maintenance ending another 5 years later.

ocp13 will require more care, as it's an important publicly available service. We can review the server specs as it might be over-compensating. We can also restore the SSH port to the default.

* [ ] ocp13: Need to replace `gateway6:` with e.g.

    ```
                  - to: default
                    via: fe80::1
    ```

### PR DESCRIPTION
**Pull request by RobHooper**
Resolves  #494 and #504.

### CODE REVIEW

**Comment by jpmckinney** [comment]
> I have noted that scrapyd needs a project configuring, on ocp13 this is configured under the datlab user (/home/datlab/.config/scrapy.cfg).

@RobHooper Is it that you observed that e.g. collect.data.open-contracting.org lists no projects? That's because we need to do something like `scrapyd-deploy kingfisher` from our local computers. Not sure why Datlab has that config.

**Comment by jpmckinney** [comment]
> I will enable mod_md again when DNS is live (sending too many failed requests, as we are at the moment, causes Lets Encrypt to block us).

Yeah, I have a note about that in the docs https://ocdsdeploy.readthedocs.io/en/latest/develop/update/apache.html

> Let’s Encrypt will reach a [Failed Validation](https://letsencrypt.org/docs/failed-validation-limit/) limit if DNS is not propagated.

I've now also added:

> In the meantime, you can [use Let’s Encrypt’s staging environment](https://ocdsdeploy.readthedocs.io/en/latest/develop/update/apache.html#mod-md-test).

**Comment by jpmckinney** [comment]
An additional manual step is to copy the `exchange_rates` table from the `pelican_backend` database.

There are docs here, for the data support server: https://ocdsdeploy.readthedocs.io/en/latest/deploy/data-support.html#pelican-backend

**Comment by jpmckinney** [comment]
In terms of migration process, is this the plan?

1. Do we adjust the TTLs to speed up the DNS switchover?
2. ocp13: I've commented out DATA_REGISTRY_CBOM so that new jobs aren't started
3. ocp27: We should re-copy any new files (`data/storage/exporter`) and databases from ocp13
1. ocp27: Can we copy the mod_md files for data.open-contracting.org from ocp13?
2. ocp27: We can deploy the server to install the deployer's crontab
3. GoDaddy: Change the DNS over for data.open-contracting.org

**Review by jpmckinney — COMMENTED** [review]
Looks good! Just one comment.

**Comment by jpmckinney** [comment]
I added a documentation page that covers most of what I wrote above (except where I still have a question). Please add the steps to restore the `data_registry` and `spoonbill` databases from backups. 

We should add a subheading on this other page, and explain in general how to restore from individual database backups versus the existing docs about pgBackrest backups: https://ocdsdeploy.readthedocs.io/en/latest/maintain/databases.html#restore-from-backup We can then just link to that from the registry page.

File: salt/aws/init.sls [hunk 1]
```diff
@@ -5,6 +5,8 @@ awscli:
     - name: python3-pip
   pip.installed:
     - name: awscli
+    - extra_args:
+      - --break-system-packages
```
[the remark this answers is not in the document]
Comment (RobHooper) [reply salt/aws/init.sls:11]: This is required on systems with Python 3.11 and newer.
I have added a comment and a Jinja if statement so it only runs where needed.

**Comment by RobHooper** [comment]
> In the meantime, you can [use Let’s Encrypt’s staging environment](https://ocdsdeploy.readthedocs.io/en/latest/develop/update/apache.html#mod-md-test).

The Lets Encrypt staging environment won't generate SSL certificates without live DNS, so this doesn't help in this situation.

> @RobHooper Did you want to [remove the disable NTP state], or you decided not to?

This was already done in a [previous PR](https://github.com/open-contracting/deploy/commit/d8e14ff2bea63cbd4cfc43f3dfe28de22570e3e1#diff-47017b3920a25904ffd32c44fb92b549751f9dcd8dbb0e9fdd2afd3b19442a70L31-L35) :)

> Do we want [modify reserved disk bytes] for the registry server?

Yes, I have ran `tune2fs -m 1 /dev/md2` now.
I think we should run this by default on any disk larger than ~50GBs.
Where would be best to document this?

**Comment by jpmckinney** [comment]
> Where would be best to document this?

Probably at an appropriate step in create_server.rst, and maybe as a note at https://ocdsdeploy.readthedocs.io/en/latest/maintain/hosting.html#rescale-a-server

**Comment by jpmckinney** [comment]
> The Lets Encrypt staging environment won't generate SSL certificates without live DNS, so this doesn't help in this situation.

It helps to not get temporarily blocked by Let's Encrypt :)

File: salt/aws/init.sls (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpmckinney) [reply salt/aws/init.sls:11]: Is this flag also needed on the pip.installed state in kingfisher/init.sls?

[the remark this answers is not in the document]
Comment (RobHooper) [reply salt/aws/init.sls:11]: We will need the above flag in kingfisher/init.sls if these packages are wanted at a system level, quickly checking them I doubt there will be any clashes / issues with them and dependencies breaking system packages.
Alternatively we can install these packages in a venv, this is better in terms of isolation but it means they are not immediately accessible.

The kingfisher/init.sls state is not ran for Registry - should it be?

**Comment by RobHooper** [comment]
Detailed migration plan as follows:

1. Reduce DNS TTLs for:

- collect.kingfisher.open-contracting.org
- downloads.kingfisher.open-contracting.org
- pelican.open-contracting.org
- postgres.kingfisher.open-contracting.org
- rabbitmq.kingfisher.open-contracting.org

2. Disable crons on ocp13:

- PELICAN_BACKEND_UPDATE_EXCHANGE_RATES
- DATA_REGISTRY_CBOM
- /etc/cron.d/postgres_backups

3. Disable sites on ocp13 (Ensure no new data is created and lost).

- systemctl stop apache2 docker
- **Enable maintenance page?**

4. Copy data from ocp13 and import on the new server

   4a. Rsync files

   ```bash
   rsync -avze "ssh -p 2223" root@ocp13.open-contracting.org:/data/storage/spoonbill/ /data/storage/spoonbill/
   rsync -avze "ssh -p 2223" root@ocp13.open-contracting.org:/data/storage/exporter_dumps/ /data/storage/exporter/
   ```

   4b. Migrate databases
   ```bash
   # Export databases on ocp13
   su - postgres -c "/usr/bin/pg_dump -Ft -f '~/spoonbill_web.tar' 'spoonbill_web'"
   su - postgres -c "/usr/bin/pg_dump -Ft -f '~/data_registry.tar' 'data_registry'"

   # Copy both files over

   # Import databases on ocp27
   sudo -u postgres pg_restore -cC --if-exists -v -U postgres -d spoonbill_web spoonbill_web.tar
   sudo -u postgres pg_restore -cC --if-exists -v -U postgres -d data_registry data_registry.tar
   ```

   4c. Migrate exchange rates table
   https://ocdsdeploy.readthedocs.io/en/latest/deploy/servers/data-support.html#pelican-backend


5. Run docker app migrations
https://ocdsdeploy.readthedocs.io/en/latest/deploy/servers/data-support.html#docker-apps

6. Test sites are working - **James will you be online for this?**

7. Update DNS to point to the ocp27 CNAME.

8. Run scrapyd-deploy once DNS has propagated - **With James?**

9. Deploy ocp27 with Salt, to install the deployer's crontab and the postgres_backups


Post-Migration
1. Copy scrapyd logs

```bash
rsync -avze "ssh -p 2223" root@ocp13.open-contracting.org:/home/collect/scrapyd/logs /ocp27/placeholder
```

2. Restore TTLs to 1 hour
3. Deploy Docs with salt updating Tinyproxy for new IP registry address
4. Deploy Prometheus with Salt
5. Block all access to ocp13, and then when we are happy decommission. (Ususally a couple of weeks after the migration).


Changes in this PR to docker and rsyslog effect all docker servers so carefully deploy and restart the Docker daemon.

**Comment by jpmckinney** [comment]
Step 0: I need to message Dogsbody to say we're decommissioning ocp13, so there will be alerts about it, data.open-contracting.org, etc.

> Reduce DNS TTLs for:

Aren't those for ocp23? We're replacing ocp13. (data.open-contracting.org, etc.)

> Disable sites on ocp13

I think we can just (1) `docker compose down` everything except the data-registry web and static containers and (2) email sysadmin@open-contracting.org that work is starting (so that Yohanna doesn't edit any publications).

> Test sites are working - James will you be online for this?

At what time are you expecting?

> Run scrapyd-deploy once DNS has propagated - With James?

Yes, I can do this one. As I understand, `DATA_REGISTRY_CBOM ` cron will be disabled until the next step.

> Copy scrapyd logs

Let's do this before step 9. Once cron starts I'm not sure how rsync'ing the directory works.

File: docs/deploy/servers/data-support.rst [hunk 2]
```diff
@@ -51,7 +51,7 @@ Docker apps
       docker compose run --rm --name django-migrate cron python manage.py migrate
 
       cd /data/deploy/pelican-frontend/
-      docker compose run --rm --name django-migrate cron python manage.py migrate
+      docker compose run --rm --name django-migrate web python manage.py migrate
```
[the remark this answers is not in the document]
Comment (jpmckinney) [reply docs/deploy/servers/data-support.rst:54]: Yup!

**Comment by jpmckinney** [comment]
Some other tasks I did:

* After disabling the CBOM cron, delete any jobs in Django admin, which cancels any jobs in Scrapyd: https://collect.data.open-contracting.org/jobs
* Check that RabbitMQ is quiet. Right now there's work being done, but it should finish by migration time: https://rabbitmq.data.open-contracting.org/#/queues

**Comment by jpmckinney** [comment]
> Changes in this PR to docker and rsyslog effect all docker servers so carefully deploy and restart the Docker daemon

Is there anything extra to do when deploying other servers, or should Docker (and containers) restart normally?

**Comment by RobHooper** [comment]
> > Changes in this PR to docker and rsyslog effect all docker servers so carefully deploy and restart the Docker daemon
> 
> Is there anything extra to do when deploying other servers, or should Docker (and containers) restart normally?

Docker should restart as part of the salt deployment causing a small outage.
I will be re-deploying on servers effected tomorrow morning while traffic is quiet.

**Pull request merged**